### PR TITLE
Multi allocator

### PR DESCRIPTION
**Pull request by lesismal**
https://github.com/lesismal/nbio/issues/130

### CODE REVIEW

File: conn_unix.go [hunk 1]
```diff
@@ -396,13 +393,15 @@ func (c *Conn) writev(in [][]byte) (int, error) {
...
-		b := mempool.Malloc(size)
+		b := c.g.writeBufferAllocator.Malloc(size)
 		copied := 0
 		for _, v := range in {
 			copy(b[copied:], v)
 			copied += len(v)
 		}
-		return c.write(b)
+		n, err := c.write(b)
+		c.g.writeBufferAllocator.Free(b)
```
Comment (acgreek) [inline conn_unix.go:403]: Is this a leak fix?

File: nbhttp/engine.go [hunk 2]
```diff
@@ -695,9 +711,9 @@ func NewEngine(conf Config) *Engine {
 			c.DataHandler(c, data)
 		}
 	})
-	g.OnWriteBufferRelease(func(c *nbio.Conn, buffer []byte) {
-		mempool.Free(buffer)
-	})
+	// g.OnWriteBufferRelease(func(c *nbio.Conn, buffer []byte) {
```
Comment (acgreek) [inline nbhttp/engine.go:714]: did you intend to leave this comment out?

File: nbhttp/processor.go [hunk 3]
```diff
@@ -37,7 +37,6 @@ func releaseRequest(req *http.Request) {
 		if req.Body != nil {
 			br := req.Body.(*BodyReader)
 			br.close()
-			bodyReaderPool.Put(br)
```
Comment (acgreek) [inline nbhttp/processor.go:40]: where are you freeing this now?

File: nbhttp/websocket/upgrader.go [hunk 4]
```diff
@@ -518,49 +519,49 @@ func (u *Upgrader) handleWsMessage(c *Conn, opcode MessageType, data []byte) {
...
+		opcode = MessageType(wr.buffer[0] & 0xF)
+		res1 = int8(wr.buffer[0]&0x40) != 0
+		res2 = int8(wr.buffer[0]&0x20) != 0
+		res3 = int8(wr.buffer[0]&0x10) != 0
+		fin = ((wr.buffer[0] & 0x80) != 0)
+		payloadLen := wr.buffer[1] & 0x7F
 		bodyLen := int64(-1)
 
 		switch payloadLen {
 		case 126:
```
Comment (acgreek) [inline nbhttp/websocket/upgrader.go:535]: good idea to define named constants for 126 and 127 with comment

File: nbhttp/websocket/upgrader.go [hunk 5]
```diff
@@ -868,9 +869,9 @@ func nextTokenOrQuoted(s string) (value string, rest string) {
 	return "", ""
 }
 
-func (u *connState) readAll(r io.Reader, size int) ([]byte, error) {
+func (wr *wsReader) readAll(r io.Reader, size int) ([]byte, error) {
 	const maxAppendSize = 1024 * 1024 * 4
```
Comment (acgreek) [inline nbhttp/websocket/upgrader.go:873]: could this result in some sort of truncation and if so, should it be configurable?

File: conn_unix.go (hunk 1 above)
Comment (lesismal) [reply conn_unix.go:403]: No, it was not a leak, we free buffer after conn.write in previous version:
https://github.com/lesismal/nbio/blob/master/conn_unix.go#L76

But in this version, who call conn.write should free the buffer itself.

File: nbhttp/engine.go (hunk 2 above)
Comment (lesismal) [reply nbhttp/engine.go:714]: will clean it

File: nbhttp/processor.go (hunk 3 above)
Comment (lesismal) [reply nbhttp/processor.go:40]: here:
https://github.com/lesismal/nbio/blob/multi_allocator/nbhttp/body.go#L83

File: nbhttp/websocket/upgrader.go (hunk 4 above)
Comment (lesismal) [reply nbhttp/websocket/upgrader.go:535]: the RFC does not define names for these different numbers of size and it's really hard to give them names. gorilla/websocket use raw number either.

File: nbhttp/websocket/upgrader.go (hunk 5 above)
Comment (lesismal) [reply nbhttp/websocket/upgrader.go:873]: sorry but I don't get the point for `configurable`, what should we config for?

**Comment by lesismal** [comment]
Too many allocators make it difficult to use and maintain, give up this time, maybe consider it again in the future.

**Pull request closed without merging**